File: libraries/authz-nestjs/src/observability/metrics.ts
import { PermissionCache } from "../permission-cache/cache";
import { CacheMode } from "../cache-sync/bootstrap";

/** Simple in-process counters/gauges (architecture §10.2). */
export class Metrics {
  private counters = new Map<string, number>();
  private gauges = new Map<string, number>();
  /** simple duration aggregates: we record _count and _sum for each base name */
  private durations = new Map<string, { count: number; sum: number }>();

  inc(name: string, by = 1): void {
    this.counters.set(name, (this.counters.get(name) ?? 0) + by);
  }

  incTokenFailure(base: string, mode: TokenFailureMode): void {
    this.inc(base);
    this.inc(`${base}_${mode}`);
  }

  setGauge(name: string, value: number): void {
    this.gauges.set(name, value);
  }

  /** Start a simple timer. Returns a stop() function that records the elapsed seconds. */
  startTimer(baseName: string): () => void {
    const start = Date.now();
    let stopped = false;
    return () => {
      if (stopped) return;
      stopped = true;
      const elapsed = (Date.now() - start) / 1000;
      this.recordDuration(baseName, elapsed);
    };
  }

  /** Record a duration (seconds) into `<base>_count` and `<base>_sum`. */
  recordDuration(baseName: string, seconds: number): void {
    const current = this.durations.get(baseName) ?? { count: 0, sum: 0 };
    current.count += 1;
    current.sum += seconds;
    this.durations.set(baseName, current);
    // expose as counters so bridge/snapshot picks them up
    this.counters.set(`${baseName}_count`, current.count);
    this.counters.set(`${baseName}_sum`, Math.round(current.sum * 1000) / 1000);
  }

  get(name: string): number {
    return this.counters.get(name) ?? this.gauges.get(name) ?? 0;
  }

  snapshot(): Record<string, number> {
    return {
      ...Object.fromEntries(this.counters),
      ...Object.fromEntries(this.gauges),
      // durations already mirrored into counters (base_count/base_sum)
    };
  }
}

export type TokenFailureMode = "expired" | "invalid_signature" | "malformed" | "other";

export function classifyTokenFailure(error: unknown): TokenFailureMode {
  const msg = error instanceof Error ? error.message.toLowerCase() : "";
  if (msg.includes("expired")) return "expired";
  if (msg.includes("signature") || msg.includes("verif")) return "invalid_signature";
  if (msg.includes("malform") || msg.includes("parse") || msg.includes("invalid jwt")) return "malformed";
  return "other";
}

export const METRIC = {
  authzSuccess: "authz_success_total",
  authzFailure: "authz_failure_total",
  permissionDenied: "authz_permission_denied_total",
  jwtValidationFailures: "jwt_validation_failures_total",
  serviceTokenFailures: "service_token_failures_total",
  serviceTokenFetchSuccess: "service_token_fetch_success_total",
  serviceTokenFetchDuration: "service_token_fetch_duration_seconds",
  roleEventSkipped: "role_event_skipped_total",
  roleRefreshFailures: "role_refresh_failures_total",
  roleRefreshSuccess: "role_refresh_success_total",
  roleEventsProcessed: "role_events_processed_total",
  roleEventProcessingDuration: "role_event_processing_duration_seconds",
  diskCacheWriteFailures: "disk_cache_write_failures_total",
  authzRequestDuration: "authz_request_duration_seconds",
} as const;

export const GAUGE = {
  cacheAgeSeconds: "permission_cache_age_seconds",
} as const;

export interface HealthReport {
  cacheStatus: "initialized" | "empty";
  cacheAgeSeconds: number;
  mode: CacheMode;
  roleServiceLastSync: string | null;
  kafkaConsumerConnected: boolean;
}

/** Build the health indicator (architecture §10.3). */
export function buildHealth(
  cache: PermissionCache,
  mode: CacheMode,
  extra?: { roleServiceLastSync?: Date | null; kafkaConsumerConnected?: boolean },
): HealthReport {
  const empty = Object.keys(cache.toSnapshot()).length === 0;
  return {
    cacheStatus: empty ? "empty" : "initialized",
    cacheAgeSeconds: Math.max(
      0,
      Math.floor((Date.now() - cache.lastUpdatedAt().getTime()) / 1000),
    ),
    mode,
    roleServiceLastSync: extra?.roleServiceLastSync
      ? extra.roleServiceLastSync.toISOString()
      : null,
    kafkaConsumerConnected: extra?.kafkaConsumerConnected ?? false,
  };
}

File: libraries/authz-spring-boot/src/main/java/com/example/authz/observability/Metrics.java
package com.example.authz.observability;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Simple in-process counters/gauges (architecture §10.2).
 *
 * <p><b>G12 — token-validation failure modes:</b> the aggregate counters
 * {@link #JWT_VALIDATION_FAILURES} and {@link #SERVICE_TOKEN_FAILURES} are
 * supplemented by per-mode counters that encode the failure cause in the
 * metric name suffix: {@code _expired}, {@code _invalid_signature},
 * {@code _malformed}. Call {@link #incTokenFailure(String, TokenFailureMode)}
 * to increment both the aggregate and the mode-specific counter atomically.
 */
public class Metrics {
    public static final String AUTHZ_SUCCESS = "authz_success_total";
    public static final String AUTHZ_FAILURE = "authz_failure_total";
    public static final String PERMISSION_DENIED = "authz_permission_denied_total";
    public static final String JWT_VALIDATION_FAILURES = "jwt_validation_failures_total";
    public static final String SERVICE_TOKEN_FAILURES = "service_token_failures_total";
    public static final String ROLE_EVENT_SKIPPED = "role_event_skipped_total";
    public static final String ROLE_REFRESH_FAILURES = "role_refresh_failures_total";
    public static final String DISK_CACHE_WRITE_FAILURES = "disk_cache_write_failures_total";
    public static final String CACHE_VERSION = "permission_cache_version";
    public static final String CACHE_AGE_SECONDS = "permission_cache_age_seconds";

    // ---- G12: per-mode failure suffixes ----

    /**
     * Token-validation failure modes (G12). Used with
     * {@link #incTokenFailure(String, TokenFailureMode)} to record both the
     * aggregate counter and a mode-specific counter
     * ({@code <base>_expired}, {@code <base>_invalid_signature},
     * {@code <base>_malformed}).
     */
    public enum TokenFailureMode {
        EXPIRED("expired"),
        INVALID_SIGNATURE("invalid_signature"),
        MALFORMED("malformed"),
        OTHER("other");

        private final String suffix;
        TokenFailureMode(String suffix) { this.suffix = suffix; }
        public String suffix() { return suffix; }
    }

    /**
     * Increments the aggregate failure counter {@code base} and the
     * mode-specific counter {@code base + "_" + mode.suffix()}.
     *
     * @param base the aggregate metric name (e.g. {@link #JWT_VALIDATION_FAILURES})
     * @param mode the failure cause
     */
    public void incTokenFailure(String base, TokenFailureMode mode) {
        inc(base);
        inc(base + "_" + mode.suffix());
    }

    /**
     * Classify a token-validation exception into a {@link TokenFailureMode}.
     * Inspects the exception message for well-known Spring Security phrases.
     */
    public static TokenFailureMode classifyTokenFailure(Throwable t) {
        if (t == null) return TokenFailureMode.OTHER;
        String msg = t.getMessage() == null ? "" : t.getMessage().toLowerCase();
        // Check cause chain too
        Throwable cause = t.getCause();
        String causeMsg = (cause != null && cause.getMessage() != null)
                ? cause.getMessage().toLowerCase() : "";
        String combined = msg + " " + causeMsg;
        if (combined.contains("expired") || combined.contains("exp")) {
            // "Jwt expired" / "token is expired"
            if (combined.contains("expir") || combined.contains("jwt expired")) {
                return TokenFailureMode.EXPIRED;
            }
        }
        if (combined.contains("expir")) return TokenFailureMode.EXPIRED;
        if (combined.contains("signature") || combined.contains("verif")) {
            return TokenFailureMode.INVALID_SIGNATURE;
        }
        if (combined.contains("malform") || combined.contains("parse")
                || combined.contains("invalid jwt") || combined.contains("bad jwt")
                || combined.contains("not a valid") || combined.contains("unable to decode")) {
            return TokenFailureMode.MALFORMED;
        }
        return TokenFailureMode.OTHER;
    }

    private final Map<String, AtomicLong> counters = new ConcurrentHashMap<>();
    private final Map<String, Long> gauges = new ConcurrentHashMap<>();

    public void inc(String name) {
        counters.computeIfAbsent(name, k -> new AtomicLong()).incrementAndGet();
    }

    public void setGauge(String name, long value) {
        gauges.put(name, value);
    }

    public long get(String name) {
        AtomicLong c = counters.get(name);
        if (c != null) return c.get();
        return gauges.getOrDefault(name, 0L);
    }

    public Map<String, Long> snapshot() {
        Map<String, Long> out = new LinkedHashMap<>();
        counters.forEach((k, v) -> out.put(k, v.get()));
        out.putAll(gauges);
        return out;
    }
}
